core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+                    possible_stringified_rules = list(
+                        cls._stringify_classified_rule(rule_spec))
+                    for possible_stringified_rule in possible_stringified_rules:
+                        if possible_stringified_rule.startswith(rule_str):
+                            return (answer_group_index, rule_spec_index, None)
+
+        return (None, None, 'Failed to match rule string')
+
+    @classmethod
+    def _infer_classification_categorization(cls, rule_str):
+        # At this point, no classification was possible. Thus, only soft, hard,
+        # and default classifications are possible.
+        fuzzy_rule_type = 'FuzzyMatches'
+        if rule_str == cls._DEFAULT_RULESPEC_STR:
+            return exp_domain.DEFAULT_OUTCOME_CLASSIFICATION
+        elif rule_str == fuzzy_rule_type:
+            return exp_domain.TRAINING_DATA_CLASSIFICATION
+        else:
+            return exp_domain.EXPLICIT_CLASSIFICATION
+
+    @classmethod
+    def _normalize_raw_answer_object(
+            cls, answer_object, raw_answer, answer_str):
+        try:
+            return (
+                answer_object.normalize(raw_answer), None)
+        except AssertionError as error:
+            answer_object_type_name = type(answer_object).__name__
+            error_str = str(error)
+            return (
+                None,
+                'Failed to normalize %s: %s (answer: \'%s\'), because of: '
+                '%s' % (
+                    answer_object_type_name, raw_answer, answer_str, error_str))
+
+    @classmethod
+    def _cb_reconstitute_code_evaluation(
+            cls, state, answer_group_index, rule_spec, rule_str, answer_str):
+        # The Jinja representation for CodeEvaluation answer strings is:
+        #   {{answer.code}}
+
+        rule_types_without_output = [
+            'CodeContains', 'CodeDoesNotContain', 'ResultsInError'
+        ]
+        # NOTE: Not all of CodeEvaluation can be reconstituted. Evaluation,
+        # error, and output (with one rule_type exception) cannot be recovered
+        # without actually running the code. For this reason, OutputEquals,
+        # CodeContains, CodeDoesNotContain, and ResultsInError can only be
+        # partially recovered. The missing values will be empty strings as
+        # special sentinel values. Empty strings must be checked in conjunction
+        # with the session_id to determine whether the empty string is the
+        # special sentinel value.
+
+        # TODO(bhenning): Should something more significant (like None) be used
+        # for sentinel values for output/evaluation/error instead?
+        if answer_str == None:
+            return (None, 'Failed to recover code: \'%s\'' % answer_str)
+        if rule_str != cls._DEFAULT_RULESPEC_STR:
+            if rule_spec.rule_type != 'OutputEquals' and (
+                    rule_spec.rule_type not in rule_types_without_output):
+                return (
```

Comment:
Yes, but we take a position of whitelisting which rules are migrated so that each type is investigated when it fails. We actually had a new rule type fail for this reason when we started migrating the larger data set of answers. It was as simple as adding the type to the array to fix it, but I think it's better to be cautious when adding new rules.